encoding/encoding.go

```diff
@@ -64,6 +78,9 @@ type StreamSerializer interface {
 	// StreamRequest creates a root stream request, a stream request reader using
 	// body reader provided
 	StreamRequest(body io.Reader) (*transport.StreamRequest, StreamRequestReader, error)
+
+	// MethodType returns the type of RPC method
+	MethodType() MethodType
```

Comment:
IMO we should add this methodic the `Encoding` interface. I understand that you only added here since only protobuf needs it. However, this method is valid to have for other encoding (in comparaison `StreamRequest` is not since it can only be used with proto)

Reply:
I'm in favour of keeping this method in `StreamSerializer` since other serializers do not support anything other than Unary which adds unnecessary boilerplate just like we found during the review initially and while only the streaming serializers have different RPC types (listed in MethodType)

Reply:
This is different than the initial review and the extra boilerplate.

In the first proposal, you needed to add extra methods for encoding that were not supported streaming. This was adding complexity, verbosity and unnecessary codes.

In this case, `MethodType` return some informations of the streaming, like `Unary`, `Bidi` etc..If you add this method to another encoding than protobuf, the method can be used and will return a valid result. `Unary` can be used for something else than just protobuf

You are actually using this method for other type of encoding, see https://github.com/yarpc/yab/pull/317/files 
For instance here:
```
func isStreamingMethod(serializer encoding.Serializer) bool {
	streamSerializer, ok := serializer.(encoding.StreamSerializer)
	if !ok {
		return false
	}

	return streamSerializer.MethodType() != encoding.Unary
```
Here we could avoid the cast to `encoding.StreamSerializer` and only use the `MethodType`.

Reply:
Non-streaming encoders are not going to have any other type of RPC method apart from Unary, IMO MethodType in streamSerializer should be okay. And I agree something meaningful can be returned from other encoders though all of them would return `Unary` (except proto-encoder)